### PR TITLE
Separate regression tests for frozen and latest checkpoints

### PR DESCRIPTION
**Pull request by mcgibbon**
This PR defines separate regression tests for "frozen" checkpoints from specific commits (which shouldn't be updated) and "latest" checkpoints ensuring ongoing backwards compatibility (which should be updated).

### CODE REVIEW

File: fme/core/registry/module.py [hunk 1]
```diff
@@ -161,7 +161,7 @@ def build(
         n_in_channels: int,
         n_out_channels: int,
         dataset_info: DatasetInfo,
-    ) -> nn.Module:
+    ) -> Module:
```
Comment (jpdunc23) [inline fme/core/registry/module.py:168]: Good catch

File: fme/core/registry/test_module_registry.py [hunk 2]
```diff
@@ -110,40 +146,48 @@ def load_or_cache_state(selector_name: str, module: Module) -> dict[str, torch.T
...
+    Backwards compatibility for frozen releases from specific commits.
+    """
+    set_seed(0)
+    module = FROZEN_BUILDERS[selector_name]()
+    loaded_state_dict = load_state(selector_name)
+    module.load_state(loaded_state_dict)
+
+
+LATEST_BUILDERS = {
+    "NoiseConditionedSFNO": get_noise_conditioned_sfno_module,
```
Comment (jpdunc23) [inline fme/core/registry/test_module_registry.py:233]: Should `NoiseConditionedSFNO_state_dict.pt` also be updated?

File: fme/core/registry/test_module_registry.py [hunk 3]
```diff
@@ -110,40 +146,48 @@ def load_or_cache_state(selector_name: str, module: Module) -> dict[str, torch.T
...
-        dataset_info=dataset_info,
-    )
+def test_frozen_module_backwards_compatibility(selector_name: str):
+    """
+    Backwards compatibility for frozen releases from specific commits.
+    """
+    set_seed(0)
+    module = FROZEN_BUILDERS[selector_name]()
+    loaded_state_dict = load_state(selector_name)
+    module.load_state(loaded_state_dict)
```
Comment (jpdunc23) [inline fme/core/registry/test_module_registry.py:229]: I'm still a bit concerned that if we place no limits on new keys then we could inadvertently introduce changes in behavior that lead to regressions in inference skill with `FROZEN_BUILDERS` checkpoints. I wonder if we should also raise an error here with a message along the lines of:

```
"New module parameters {new_keys} found that were not present in the 
"frozen" checkpoint. New module parameters may be added but should 
be enabled by adding a new config parameter that, by default, does not 
add the new parameters when building the module."
```

We could also save and reload the module config dict together with the artifact to verify that the config builds the same architecture. Of course there are a million other ways to change the module code that could lead to inference regressions, but I don't see why we should allow arbitrary new parameters that weren't present when the checkpoint was saved if there is a way to avoid it.

**Review by jpdunc23 — COMMENTED** [review]
I'm still concerned about silently allowing new parameters with frozen checkpoints, though it maybe warrants a discussion at a tech sync and I'm willing to punt for now if you prefer.

Other than that, I have one blocking comment about updating the existing artifact.

File: fme/core/registry/test_module_registry.py (hunk 2 above)
Comment (mcgibbon) [reply fme/core/registry/test_module_registry.py:233]: It was already a "latest" version.

File: fme/core/registry/test_module_registry.py (hunk 3 above)
Comment (mcgibbon) [reply fme/core/registry/test_module_registry.py:229]: That type of thing is supposed to be covered by the "produces the same result" test(s). If you're concerned about new parameters not affecting an initial prediction but affecting later ones after the first gradient update, I should add a second stage to those tests that does a second step when testing for identicality.

I see though now what you're saying, I've not been understanding it. In practice, what I have here won't catch any of the cases I care about updating the regression tests for, because we always add them in a way that sets the weights to None (which doesn't get registered in the state dict). Really what we need is to remember to update/write a new test when we add features that define new weights.

What I actually want to do is test that the _config_ has no new keys, and force the user to build a new latest checkpoint when new config keys are added, saving the asdict'd config with the checkpoint. I'll see about adding that, and also adding what you suggested about making sure the config builds the same architecture.

Comment (mcgibbon) [reply fme/core/registry/test_module_registry.py:229]: I looked at it more, and because we load model states with `strict=True` (the default), this will already error if the built model keys differ from the checkpoint. I can see how my other (wrong) test implied this wasn't the case, though.

File: fme/core/registry/test_module_registry.py [hunk 4]
```diff
@@ -140,10 +144,121 @@ def test_module_backwards_compatibility(selector_name: str):
...
+    if not state_dict_path.exists():
+        raise RuntimeError(
+            f"State dict for {selector_name} not found at {state_dict_path}. "
+            "Please make sure the checkpoint exists and is committed to the repo."
+        )
+    return torch.load(state_dict_path)
+
+
+def load_or_cache_state(
+    selector_name: str, module: Module, module_config: ModuleConfig | None = None
```
Comment (jpdunc23) [inline fme/core/registry/test_module_registry.py:180]: The `module_config` argument to `load_or_cache_state` appears to be unused.

File: fme/core/registry/test_module_registry.py [hunk 5]
```diff
@@ -79,8 +82,27 @@ def test_module_selector_raises_with_bad_config():
...
+    )
+    vertical_coordinate = HybridSigmaPressureCoordinate(
+        ak=torch.arange(7, device=device), bk=torch.arange(7, device=device)
+    )
+    dataset_info = DatasetInfo(
+        horizontal_coordinates=horizontal_coordinate,
+        vertical_coordinate=vertical_coordinate,
+        timestep=timestep,
+        all_labels=all_labels,
+    )
```
Comment (jpdunc23) [inline fme/core/registry/test_module_registry.py:104]: Consider adding a new / reusing an existing shared helper to create the `DatasetInfo` that both `get_dbc2925_ncsfno_module` and `get_noise_conditioned_sfno_module` can reuse. 

I think it makes sense to repeat the `select = ModuleSelector(` blocks, even if they're currently identical.

File: fme/core/registry/test_module_registry.py (hunk 3 above)
Comment (jpdunc23) [reply fme/core/registry/test_module_registry.py:229]: Ah, good point. Maybe we should add a test that confirms this behavior for `Module.load_state` (and maybe also other parts of `Module`). But that's a preexisting issue.
